Operational-Procedures/Scheduler/troubleshooting-the-scheduling-algorithm/index.rst

```diff
@@ -0,0 +1,177 @@
+
+#.  Install the rubin-sim package with mamba: 
+
+    .. prompt:: bash
+
+        mamba install -c conda-forge -c lsstts rubin-sim
+
+    The rubin-sim package contains several modules for simulating and analysing Rubin data, and includes the feature based scheduler module, we will need to inspect the snapshots.
+
+#.  Create a notebook environment with the scheduler-dev conda environment:
+
+    .. prompt:: bash
+
+        ipython kernel install --name "scheduler-dev" --user
+    
+    This will allow us to select the "scheduler-dev" environment for our Jupyter notebooks.
+
+After the steps above are completed, create a new notebook.
+On the top-right-hand side, click on the environment name, where it is says "LSST" and follow the sequence below to select the "scheduler-dev" environment.
+
+.. list-table:: 
+
+    * - .. figure:: ./_static/new-jupyter-notebook-lsst.png
+    * - .. figure:: ./_static/new-jupyter-notebook-lsst-selected.png
+    * - .. figure:: ./_static/new-jupyter-notebook-lsst-select-kernel.png
+    * - .. figure:: ./_static/new-jupyter-notebook-scheduler-dev-selection.png
+    * - .. figure:: ./_static/new-jupyter-notebook-scheduler-dev.png
+
+You are now ready to analyse some Scheduler snapshots.
+
+.. note::
+
+    The environment we created to analyse the scheduler snapshots does not have the libraries required to access the control system.
+    Keep that in mind for the following procedures.
+
+.. _troubleshooting-the-scheduling-algorithm-retrieving-snapshots:
+
+Retrieving Snapshots
+====================
+
+To retrieve a snapshot created by Scheduler, first we need to find its ``url``.
+For that, follow the :ref:`scheduler-night-time-operation-troubleshooting-recovering-from-a-scheduler-fault-find-last-scheduler-snapshot` procedure.
+
+Once you have the ``url`` you can retrieve the snapshot directly from a notebook.
+
+We will use this same notebook to analyse the data, so we start by loading all the libraries we will need.
+
+.. code-block:: python
+
+    import io
+    import os
+    import math
+    import pickle
+    import importlib
+
+    import urllib.request
+
+    import numpy as np
+    import healpy as hp
+
+In the cell below we use the same ``url`` found in :ref:`scheduler-night-time-operation-troubleshooting-recovering-from-a-scheduler-fault-find-last-scheduler-snapshot`, make sure to update the entire string with the appropriate value.
+
+.. code-block:: python
+
+    uri = "https://s3.cp.lsst.org/rubinobs-lfa-cp/Scheduler:2/Scheduler:2/2022/02/17/Scheduler:2_Scheduler:2_2022-02-18T09:26:04.347.p"
+
+Now we use ``urlretrieve`` to download the file and save it to a local file named ``scheduler_snapshot.p``.
+
+.. code-block:: python
+
+    dest, _ = urllib.request.urlretrieve(url=uri, filename="./scheduler_snapshot.p")
+
+Finally, we load the content of the snapshot.
+
+.. code-block:: python
+
+    with open(dest, "rb") as fp:
+        scheduler, conditions = pickle.load(fp)
+
```

Comment:
Following these steps I got the following error traceback when attempting the pickel.load(fp) 
`INFO:numexpr.utils:Note: NumExpr detected 64 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
INFO:numexpr.utils:NumExpr defaulting to 8 threads.
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [6], in <cell line: 1>()
      1 with open(dest, "rb") as fp:
----> 2     scheduler, conditions = pickle.load(fp)

AttributeError: Can't get attribute 'BalanceVisits' on <module 'rubin_sim.scheduler.basis_functions.basis_functions' from '/home/edennihy/.conda/envs/scheduler-dev/lib/python3.8/site-packages/rubin_sim/scheduler/basis_functions/basis_functions.py'>
`

Reply:
I updated the documentation on how to install `rubin-sim`.. The `lsstts` channel has to be before `conda-forge`...